### PR TITLE
Datastore integration

### PR DESCRIPTION
**Pull request by dignifiedquire**
Dragons inside ! 🐉 🐉 🐉 

Integration with the new datastores, will have breaking api changes! 
Will fix https://github.com/ipfs/js-ipfs/issues/229 due to using https://github.com/dignifiedquire/lock-me for aquiring file system locks and exposing need methods to read and write the http api addr file.

ref https://github.com/ipfs/js-ipfs/issues/787

- [x] locking working in browser and node
- [x] expose new init function
- [x] disable flatfs sharding in the browser
- [x] flatfs compatible with the latest go-ipfs
- [x] blockstore interop with go
- [x] read and write config
- [x] read and write repo version
- [x] datastore interop 
- [x] datastore tests

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,56 +1,246 @@
...
+  /**
+   * Close the repo and cleanup.
+   *
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  close (callback) {
+    if (this.closed) {
+      return callback(new Error('repo is already closed'))
     }
```
Comment (diasdavid) [inline src/index.js]: Different behaviour. open will return silently if it is already open, but close will return an error. Or return an error in both or silent in both.

File: src/index.js [hunk 2]
```diff
@@ -1,56 +1,246 @@
...
+   * @returns {void}
+   */
+  getPrivateKey (callback) {
+    this.config.get((err, config) => {
+      if (err) {
+        return callback(err)
+      }
+      callback(null, config.Identity.PrivKey)
+    })
+  }
```
Comment (diasdavid) [inline src/index.js]: Why do we need this method here? Repo should not assume a specific format of the config file.

File: README.md [hunk 3]
```diff
@@ -45,62 +34,78 @@ This is the implementation of the [IPFS repo spec](https://github.com/ipfs/specs
...
-more. Each of the individual repos has an interface defined by
-[interface-pull-blob-store](https://github.com/ipfs/interface-pull-blob-store): this
-enables us to make IPFS Repo portable (running on Node.js vs the browser) and
-accept different types of storage mechanisms for each repo (fs, levelDB, etc).
-
-### Good to know (historical context)
-
-- The datastore folder holds the legacy version of datastore, still built in levelDB, there is a current endeavour of pushing it to fs completely.
-- The blocks folder is the current version of datastore.
-- The keys repo doesn't exist yet, as the private key is simply stored inside config
```
Comment (dignifiedquire) [inline README.md]: It's not correct anymore, that's why I removed it for now

File: example.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+'use strict'
```
Comment (dignifiedquire) [inline example.js]: no, this is used in the docs generation `example.js` is a special file name

File: package.json [hunk 5]
```diff
@@ -3,24 +3,31 @@
   "version": "0.11.3",
   "description": "IPFS Repo implementation",
   "main": "src/index.js",
+  "browser": {
+    "./src/lock.js": "./src/lock-browser.js",
+    "./src/default-config.js": "./src/default-config-browser.js"
+  },
```
Comment (dignifiedquire) [inline package.json]: fair point, any thoughts on better naming?

File: src/index.js (hunk 2 above)
Comment (dignifiedquire) [inline src/index.js]: This was in here before, I just migrated the functionality

File: README.md [hunk 6]
```diff
@@ -45,62 +34,78 @@ This is the implementation of the [IPFS repo spec](https://github.com/ipfs/specs
...
+         ┌───────────┴───────────┐                           ┌───────────┴───────────┐
+         │                       │                           │                       │
+         │                       │                           │                       │
+┌─────────────────┐     ┌─────────────────┐         ┌─────────────────┐     ┌─────────────────┐
+│                 │     │                 │         │                 │     │                 │
+│     /blocks     │     │   /datastore    │         │     /blocks     │     │   /datastore    │
+│                 │     │                 │         │                 │     │                 │
+├─────────────────┤     ├─────────────────┤         ├─────────────────┤     ├─────────────────┤
+│ FlatfsDatastore │     │LevelDBDatastore │         │LevelJSDatastore │     │LevelJSDatastore │
+└─────────────────┘     └─────────────────┘         └─────────────────┘     └─────────────────┘
```
Comment (dignifiedquire) [inline README.md]: it's a drawing of the datastores and how the are used, what you had before were not actual stores

File: src/index.js (hunk 2 above)
Comment (diasdavid) [inline src/index.js]: ␡💥

File: package.json [hunk 7]
```diff
@@ -31,27 +38,29 @@
     "node": ">=4.0.0"
```
Comment (dignifiedquire) [inline package.json]: done

File: src/index.js (hunk 1 above)
Comment (dignifiedquire) [inline src/index.js]: done, will always error now

File: src/index.js (hunk 2 above)
Comment (dignifiedquire) [inline src/index.js]: done

File: package.json (hunk 5 above)
Comment (dignifiedquire) [inline package.json]: renamed to `default-options`

File: README.md [hunk 8]
```diff
@@ -119,77 +124,26 @@ Loading this module through a script tag will make the `IpfsRepo` obj available
...
-var inMemoryBS = require('interface-pull-blob-store')
-// inMemoryBS is an "in memory" blob store, you can find others at:
-// https://github.com/ipfs/interface-pull-blob-store#modules-that-use-this
-
-var IPFSRepo = require('ipfs-repo')
-var repo = new IPFSRepo('/Users/someone/.ipfs', {
-  stores: inMemoryBS
+const Repo = require('ipfs-repo')
+const repo = new Repo('/Users/awesome/.jsipfs')
+repo.init(defaultConfig, (err) => {
```
Comment (dignifiedquire) [inline README.md]: fixed

File: README.md [hunk 9]
```diff
@@ -119,77 +124,26 @@ Loading this module through a script tag will make the `IpfsRepo` obj available
...
 ```js
-var inMemoryBS = require('interface-pull-blob-store')
-// inMemoryBS is an "in memory" blob store, you can find others at:
-// https://github.com/ipfs/interface-pull-blob-store#modules-that-use-this
-
-var IPFSRepo = require('ipfs-repo')
-var repo = new IPFSRepo('/Users/someone/.ipfs', {
-  stores: inMemoryBS
+const Repo = require('ipfs-repo')
+const repo = new Repo('/Users/awesome/.jsipfs')
```
Comment (dignifiedquire) [inline README.md]: done

File: README.md (hunk 3 above)
Comment (diasdavid) [inline README.md]: Only the last item is not valid anymore

File: example.js (hunk 4 above)
Comment (diasdavid) [inline example.js]: I see. I'm not very keen of putting even more stuff at the root level of each repo.

File: README.md [hunk 10]
```diff
@@ -119,77 +124,26 @@ Loading this module through a script tag will make the `IpfsRepo` obj available
...
+    throw err
+  }
+
+  repo.open((err) => {
+    if (err) {
+      throw err
+    }
+
+    console.log('repo is ready')
+  })
```
Comment (diasdavid) [inline README.md]: 🛎

File: example.js (hunk 4 above)
Comment (dignifiedquire) [inline example.js]: we already have it in many repos, so what do you want to do about it?

File: README.md (hunk 3 above)
Comment (dignifiedquire) [inline README.md]: the first item is also not true, datastore is not legacy, it's being actively used and things added to it

Comment (dignifiedquire) [inline README.md]: which makes the second incorrect as well

Comment (dignifiedquire) [inline README.md]: making the last item the only one valid

File: README.md (hunk 10 above)
Comment (dignifiedquire) [inline README.md]: done

File: example.js (hunk 4 above)
Comment (diasdavid) [inline example.js]: Nothing now, let's revisit docs strategy and execution for next quarter

File: package.json (hunk 5 above)
Comment (diasdavid) [inline package.json]: thank you :)

**Pull request merged**